keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,437 @@
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place
+- [ ] (R) Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [etcd-io/website], for publication to [etcd.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[etcd.io]: https://etcd.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[etcd-io/website]: https://github.com/etcd-io/website
+
+## Summary
+
+We are introducing the Kubernetes style feature gate framework into etcd to gate future feature enhancement behind a sequence of feature lifecyles. Users would be able to turn features on or off and query feature enablement in a consistent way. 
+
+## Motivation
+
+Currently any new enhancements to the etcd are typically added as [an experimental feature](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#adding-a-new-feature), with a configuration flag prefixed with "experimental", e.g. --experimental-feature-name. 
+
+When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.
+
+We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.
+
+Benefits from this enhancement includes:
+
+* graduating a feature is no longer a breaking change with new flags.
+
+* well-established feature lifecycle convention from Kubernetes.
+
+* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.
+
+* easy to query if a feature is available for an ectd cluster.
+
+### Goals
+
+* Introduce the `--feature-gates` flag in etcd code base.
+
+* Introduce apis/methods to query the enablement state of a feature.
+
+* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.
```

Comment:
Since we also cover consensus-based feature enablement in this KEP, can we add that as an explicit goal here? Say `Establish a mechanism to safely and consistently enable/disable cluster-level features`

Reply:
Thanks @shyamjvs. These are all very good points. After some thoughts, we decided to move cluster level feature gate into another KEP, we can continue the discussions there https://github.com/kubernetes/enhancements/issues/4647

Reply:
+1 to that - will surely make this KEP more tractable.

One question - is there any coupling we expect between member-level and cluster-level features (like the APIs/mechanisms to enable and expose them)? If so, it'll help to roughly address in this KEP how we're leaving space to address the latter usecase in future.

Reply:
In this KEP, there is a line of `we will not expose querying feature enablement of any cluster level features, because we cannot risk possible conflicting settings that might break cluster correctness.`. 
The alternative is to expose the cluster level features as per server feature, and users can aggregate them across all members, but that could be prone to misusage like when to use aggregation, and not really based on a raft consensus.

Reply:
yeah verifiability and correctness will be hard to prove from client-side imho (similar to how `/members` API gives consistent list of members rather than clients assuming it to be "running instances")

Reply:
> The alternative is to expose the cluster level features as per server feature, and users can aggregate them across all members, but that could be prone to misusage like when to use aggregation, and not really based on a raft consensus.

Don't. Let's follow up the discussion when we have a KEP to propose a cluster level feature gates.